File: js/src/provider.ts
import {
    resourceFromAttributes,
    type Resource,
} from "@opentelemetry/resources"
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto"
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base"
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node"
import type { NetraConfig } from "./config.js"

export function buildResource(cfg: NetraConfig): Resource {
    const attrs: Record<string, string> = {
        "service.name": cfg.project ?? "netra-observe",
    }
    if (cfg.project) attrs["netra.project"] = cfg.project
    if (cfg.environment) attrs["deployment.environment"] = cfg.environment
    return resourceFromAttributes(attrs)
}

export function buildExporter(cfg: NetraConfig): OTLPTraceExporter {
    return new OTLPTraceExporter({
        url: `${cfg.endpoint}/v1/traces`,
        headers: { Authorization: `Bearer ${cfg.apiKey}` },
    })
}

export function buildProcessor(cfg: NetraConfig): BatchSpanProcessor {
    return new BatchSpanProcessor(buildExporter(cfg))
}

export function buildProvider(cfg: NetraConfig): NodeTracerProvider {
    return new NodeTracerProvider({
        resource: buildResource(cfg),
        spanProcessors: [buildProcessor(cfg)],
    })
}

File: python/src/netra_observe/_provider.py
from __future__ import annotations

from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from ._config import Config


def build_resource(cfg: Config) -> Resource:
    attrs = {"service.name": cfg.project or "netra-observe"}
    if cfg.project:
        attrs["netra.project"] = cfg.project
    if cfg.agent:
        attrs["gen_ai.agent.name"] = cfg.agent
    if cfg.environment:
        attrs["deployment.environment"] = cfg.environment
    return Resource.create(attrs)


def build_exporter(cfg: Config) -> OTLPSpanExporter:
    return OTLPSpanExporter(
        endpoint=cfg.endpoint + "/v1/traces",
        headers={"Authorization": f"Bearer {cfg.api_key}"},
    )


def build_processor(cfg: Config) -> BatchSpanProcessor:
    return BatchSpanProcessor(build_exporter(cfg))


def build_provider(cfg: Config) -> TracerProvider:
    provider = TracerProvider(resource=build_resource(cfg))
    provider.add_span_processor(build_processor(cfg))
    return provider
